sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/Counter.java

```diff
@@ -295,6 +296,56 @@ public AggregationKind getKind() {
   public abstract CounterMean<T> getMean();
 
   /**
+   * CommitState represents whether counters' values have been committed to the backend.
+   *
+   * <p>Runners can use this information to optimize counters updates.
+   * For example, if counters are committed, runners may choose to skip the updates.
+   *
+   * <p>Counters' state transition table:
+   * Action\Current State         COMMITTED        DIRTY        COMMITTING
+   * add()                        DIRTY            DIRTY        DIRTY
+   * committing()                 None             COMMITTING   None
+   * committed()                  None             None         COMMITTED
+   */
+  @VisibleForTesting
+  enum CommitState {
+    COMMITTED,
+    DIRTY,
+    COMMITTING,
+  }
+
+  /**
+   * Returns if the counter contains non-committed aggregate.
+   */
+  public boolean isDirty() {
+    return commitState.get() != CommitState.COMMITTED;
+  }
+
+  /**
+   * Changes the counter from {@code CommitState.DIRTY} to {@code CommitState.COMMITTING}.
+   */
+  public boolean committing() {
+    return commitState.compareAndSet(CommitState.DIRTY, CommitState.COMMITTING);
+  }
+
+  /**
+   * Changes the counter from {@code CommitState.COMMITTING} to {@code CommitState.COMMITTED}.
+   */
+  public boolean committed() {
+    return commitState.compareAndSet(CommitState.COMMITTING, CommitState.COMMITTED);
+  }
+
+  /**
+   * Sets the counter to {@code CommitState.DIRTY}.
+   *
+   * <p>It needs to be called at the beginning of {@link #addValue}, {@link #resetToValue},
```

Comment:
Ultimately:

Document the two requirements for a user of Counter to meet for proper behavior

1. After all possible mutations have completed, the reader should check every counter's DIRTY state, otherwise updates may be lost
2. A counter may become DIRTY without a corresponding update to the value. This generally will occur when the calls to `addValue()`, `committing()`, and `committed()` are interleaved such that the value is updated between the calls to committing and the read of the value.